// @linthtml/linthtml/dist/rules/tag-self-close/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const dom_utils_1 = require("@linthtml/dom-utils");
const validate_option_1 = require("../../validate_option");
const RULE_NAME = "tag-self-close";
function lint(node, style, { report }) {
    if (!(0, dom_utils_1.is_tag_node)(node) || (0, dom_utils_1.is_void_node)(node) === false) {
        return;
    }
    const { name, close } = node;
    // If the tag did not close itself
    // remove toLowerCase
    if (!close || name.toLowerCase() !== close.chars.toLowerCase()) {
        const selfClose = (0, dom_utils_1.is_self_closing)(node);
        if ((style === "always" && !selfClose) || (style === "never" && selfClose)) {
            report({
                code: "E018",
                position: node.open.loc,
                meta: {
                    data: {
                        expect: style
                    }
                }
            });
        }
    }
}
exports.default = {
    name: RULE_NAME,
    validateConfig: (0, validate_option_1.create_list_value_validator)(RULE_NAME, ["always", "never"], false),
    lint
};
