// @linthtml/linthtml/dist/rules/class-no-dup/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const dom_utils_1 = require("@linthtml/dom-utils");
const util_1 = require("util");
const { isRegExp } = util_1.types;
const RULE_NAME = "class-no-dup";
function filterClasses(classes, options) {
    let ignore = options["id-class-ignore-regex"];
    // TODO: Remove after `raw-ignore-text` refacto
    classes = classes.filter((_) => /^¤+$/.test(_) === false);
    if (ignore) {
        ignore = isRegExp(ignore) ? ignore : new RegExp(ignore);
        classes = classes.filter((_class) => !ignore.test(_class));
    }
    return classes;
}
function lint(node, _config, { report, global_config }) {
    if ((0, dom_utils_1.is_tag_node)(node) && (0, dom_utils_1.has_attribute)(node, "class")) {
        const class_attribute = (0, dom_utils_1.get_attribute)(node, "class");
        let classes = filterClasses((0, dom_utils_1.get_classes)(class_attribute), global_config);
        classes = classes.sort();
        for (let i = 0; i < classes.length - 1; i++) {
            if (classes[i + 1] === classes[i]) {
                report({
                    code: "E041",
                    position: class_attribute.value.loc,
                    meta: {
                        data: { classes: classes[i] }
                    }
                });
            }
        }
    }
}
exports.default = {
    name: RULE_NAME,
    lint
};
