// @linthtml/linthtml/dist/rules/id-style/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const check_format_1 = tslib_1.__importDefault(require("../../utils/check_format"));
const dom_utils_1 = require("@linthtml/dom-utils");
const validate_option_1 = require("../../validate_option");
const util_1 = require("util");
const { isRegExp } = util_1.types;
const RULE_NAME = "id-style";
// TODO: Check behavoir with empty id attribute
function lint(node, format, { report, global_config }) {
    if ((0, dom_utils_1.is_tag_node)(node) && format !== "none") {
        const ignore = global_config["id-class-ignore-regex"];
        // TODO: Remove after `raw-ignore-text` refacto
        let attributes = node.attributes
            .filter(({ name }) => /^¤+$/.test(name.chars) === false)
            .filter(({ name }) => name.chars.toLowerCase() === "id")
            .filter(({ value }) => value && /^¤+$/.test(value.raw) === false);
        if (ignore) {
            const R = isRegExp(ignore) ? ignore : new RegExp(ignore);
            attributes = attributes.filter(({ value }) => value && R.test(value.chars) === false); // raw or chars ?
        }
        attributes.forEach((attribute) => {
            var _a;
            const id = (_a = attribute.value) === null || _a === void 0 ? void 0 : _a.chars;
            if ((0, check_format_1.default)(format, id) === false) {
                report({
                    code: "E011",
                    position: attribute.loc,
                    meta: {
                        data: {
                            attribute: "id",
                            format: format,
                            value: id
                        }
                    }
                });
            }
        });
    }
}
exports.default = {
    name: RULE_NAME,
    validateConfig: (0, validate_option_1.create_list_value_validator)(RULE_NAME, ["none", "lowercase", "underscore", "dash", "camel", "bem"]),
    lint
};
