// @linthtml/linthtml/dist/rules/attr-no-unsafe-char/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const dom_utils_1 = require("@linthtml/dom-utils");
const RULE_NAME = "attr-no-unsafe-char";
const regUnsafe = 
/* eslint-disable-next-line no-control-regex, no-misleading-character-class */
/[\u0000-\u0009\u000b\u000c\u000e-\u001f\u007f-\u009f\u00ad\u0600-\u0604\u070f\u17b4\u17b5\u200c-\u200f\u2028-\u202f\u2060-\u206f\ufeff\ufff0-\uffff]/;
function lint(node, _config, { report }) {
    if ((0, dom_utils_1.is_tag_node)(node)) {
        // TODO: Remove after `raw-ignore-text` refacto
        const attributes = node.attributes.filter(({ name }) => /^¤+$/.test(name.chars) === false);
        attributes.forEach((attribute) => {
            if ((0, dom_utils_1.attribute_has_value)(node, attribute.name.chars, regUnsafe)) {
                report({
                    code: "E004",
                    position: attribute.loc,
                    meta: {
                        data: {
                            attribute: attribute.name.chars
                        }
                    }
                });
            }
        });
    }
}
exports.default = {
    name: RULE_NAME,
    lint
};
