// @linthtml/linthtml/dist/rules/link-min-length-4/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const dom_utils_1 = require("@linthtml/dom-utils");
const RULE_NAME = "link-min-length-4";
function get_text_content(node) {
    if ((0, dom_utils_1.is_comment_node)(node)) {
        return "";
    }
    if ((0, dom_utils_1.is_text_node)(node)) {
        return node.data;
    }
    return node.children.reduce((content, child) => `${content}${get_text_content(child)}`, "");
}
function lint(node, _config, { report }) {
    var _a, _b;
    // Will add a rule to enforce href attribute on link
    if ((0, dom_utils_1.is_tag_node)(node) && node.name === "a" && (0, dom_utils_1.has_non_empty_attribute)(node, "href")) {
        const content = get_text_content(node).trim();
        const aria_label = (_b = (_a = (0, dom_utils_1.attribute_value)(node, "aria-label")) === null || _a === void 0 ? void 0 : _a.chars) !== null && _b !== void 0 ? _b : "";
        // TODO: need to deal with aria-labelledby
        if (content.length < 4 && aria_label.length < 4) {
            report({
                code: "E059",
                position: node.open.loc,
                meta: {
                    data: {
                        content
                    }
                }
            });
        }
    }
}
exports.default = {
    name: RULE_NAME,
    lint
};
