// @linthtml/linthtml/dist/rules/title-max-len/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const dom_utils_1 = require("@linthtml/dom-utils");
const validate_option_1 = require("../../validate_option");
const RULE_NAME = "title-max-len";
// TODO: check what happen if empty config/add default value?
function lint(node, max_length, { report }) {
    if ((0, dom_utils_1.is_tag_node)(node) && node.name === "title") {
        const text = node.children
            .filter((child) => (0, dom_utils_1.is_text_node)(child))
            .map((child) => child.data)
            .join("");
        if (text.length > max_length) {
            report({
                code: "E029",
                position: node.open.loc,
                meta: {
                    data: {
                        title: text,
                        max_lengthgth: max_length
                    }
                }
            });
        }
    }
}
exports.default = {
    name: RULE_NAME,
    validateConfig: (0, validate_option_1.create_number_validator)(RULE_NAME, false),
    lint
};
