// @linthtml/linthtml/dist/rules/attr-quote-style/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const dom_utils_1 = require("@linthtml/dom-utils");
const validate_option_1 = require("../../validate_option");
const RULE_NAME = "attr-quote-style";
const formats = {
    double: { regex: /^"/, desc: "double quoted" },
    single: { regex: /^'/, desc: "single quoted" },
    quoted: { regex: /^['"]/, desc: "quoted" }
};
function lint(node, config, { report }) {
    if ((0, dom_utils_1.is_tag_node)(node)) {
        // TODO: Remove after `raw-ignore-text` refacto
        const attributes = node.attributes.filter(({ name }) => /^¤+$/.test(name.chars) === false);
        const format = formats[config];
        attributes.forEach((attribute) => {
            if (attribute.value !== null && !format.regex.test(attribute.value.raw)) {
                report({
                    code: "E005",
                    position: attribute.value.loc,
                    meta: {
                        data: {
                            attribute: attribute.name.chars,
                            format: format.desc
                        }
                    }
                });
            }
        });
    }
}
exports.default = {
    name: RULE_NAME,
    validateConfig: (0, validate_option_1.create_list_value_validator)(RULE_NAME, ["double", "single", "quoted"], false),
    lint
};
