// @linthtml/linthtml/dist/rules/tag-req-attr/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const dom_utils_1 = require("@linthtml/dom-utils");
const RULE_NAME = "tag-req-attr";
function validateConfig(options) {
    if (typeof options !== "object") {
        throw new Error(`Configuration for rule "${RULE_NAME}" is invalid: Expected object got ${typeof options}`);
    }
    return options;
}
// TODO: Cleanup config validation
function lint(node, tags, { report }) {
    if ((0, dom_utils_1.is_tag_node)(node)) {
        for (const tagName in tags) {
            if (tagName === node.name) {
                const requiredAttributes = tags[tagName];
                requiredAttributes.forEach(({ name, allowEmpty }) => {
                    allowEmpty = typeof allowEmpty === "undefined" ? false : allowEmpty;
                    if (!(0, dom_utils_1.has_attribute)(node, name) || !(0, dom_utils_1.has_non_empty_attribute)(node, name, allowEmpty)) {
                        report({
                            code: "E057",
                            position: node.open.loc,
                            meta: {
                                data: {
                                    attribute: name,
                                    tag: node.name
                                }
                            }
                        });
                    }
                });
            }
        }
    }
}
exports.default = {
    name: RULE_NAME,
    validateConfig,
    lint
};
