// @linthtml/linthtml/dist/rules/attr-req-value/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const dom_utils_1 = require("@linthtml/dom-utils");
const RULE_NAME = "attr-req-value";
function lint(node, _config, { report }) {
    if ((0, dom_utils_1.is_tag_node)(node)) {
        // TODO: Remove after `raw-ignore-text` refacto
        const attributes = node.attributes.filter(({ name }) => /^¤+$/.test(name.chars) === false);
        attributes.forEach((attribute) => {
            const name = attribute.name.chars.toLowerCase();
            if (!(0, dom_utils_1.has_non_empty_attribute)(node, name) && !(0, dom_utils_1.is_boolean_attribute)(attribute)) {
                report({
                    code: "E006",
                    position: attribute.loc,
                    meta: {
                        data: {
                            attribute: name
                        }
                    }
                });
            }
        });
    }
}
exports.default = {
    name: RULE_NAME,
    lint
};
