// @linthtml/linthtml/dist/rules/id-class-no-ad/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const dom_utils_1 = require("@linthtml/dom-utils");
const RULE_NAME = "id-class-no-ad";
function lint(node, _config, { report }) {
    if ((0, dom_utils_1.is_tag_node)(node)) {
        // TODO: Remove after `raw-ignore-text` refacto
        const attributes = node.attributes
            .filter(({ name }) => /^¤+$/.test(name.chars) === false)
            .filter(({ name, value }) => ["id", "class"].indexOf(name.chars) !== -1 && value !== null);
        // TODO: Check what happen with empty id/class attributes
        attributes.forEach((attribute) => {
            var _a;
            const regex = /(^|[^a-zA-Z0-9])(ad|banner|social)(?![a-zA-Z0-9])/;
            const match = regex.exec((_a = attribute === null || attribute === void 0 ? void 0 : attribute.value) === null || _a === void 0 ? void 0 : _a.chars);
            if (match) {
                report({
                    code: "E010",
                    position: attribute.loc,
                    meta: {
                        data: {
                            word: match[2],
                            attribute: attribute.name.chars
                        }
                    }
                });
            }
        });
    }
}
exports.default = {
    name: RULE_NAME,
    lint
};
