// @linthtml/linthtml/dist/rules/focusable-tabindex-style/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const dom_utils_1 = require("@linthtml/dom-utils");
const RULE_NAME = "focusable-tabindex-style";
function is_linkable(node) {
    const isLink = ["a", "area"].indexOf(node.name) !== -1;
    return isLink ? (0, dom_utils_1.has_attribute)(node, "href") : false;
}
function is_focusable_node(node) {
    const isFocusableNatively = ["button", "input", "select", "textarea"].indexOf(node.name) !== -1;
    return isFocusableNatively || is_linkable(node) || (0, dom_utils_1.has_non_empty_attribute)(node, "tabindex");
}
function lint(node, _config, { report }) {
    if ((0, dom_utils_1.is_tag_node)(node) && is_focusable_node(node) && (0, dom_utils_1.has_attribute)(node, "disabled") === false) {
        const tabindex = (0, dom_utils_1.attribute_value)(node, "tabindex");
        if (tabindex && parseInt(tabindex.chars, 10) > 0) {
            report({
                code: "E026",
                position: node.open.loc,
                meta: {
                    data: {
                        tabindex: parseInt(tabindex.chars, 10)
                    }
                }
            });
        }
    }
}
exports.default = {
    name: RULE_NAME,
    lint
};
